### PR TITLE
Team2-tund20-implement testcases about search location module

### PR DESCRIPTION
**Pull request by tund-hcmue**
Update WebKeyword:
- keyword setValueForSelectElement
- keyword click
- add new method waitForElementToBeClickable
Generate extent report:
- Create new folder utils, extentreport, listener and reports
- Create new class ExtentManager, ExtentTestManager, ReportListener
Implement search location module
- Create new class CarForSaleSearchLocationPage, CarForSaleSearchLocationTest

### CODE REVIEW

File: demo-automation-java/src/test/java/com/example/pages/BasePage.java [hunk 1]
```diff
@@ -39,7 +39,28 @@ public void navigateToPage(String menuText) {
...
+ * @return true or false
+ */
+public boolean isElementDisplayed(WebElement webElement){
+    if (webElement.isDisplayed()){
+        return true;
+    }else return false;
+}
+
+/**
+ * Method for verify that element shoulde be enable
```
Comment (TrangDao25399) [inline demo-automation-java/src/test/java/com/example/pages/BasePage.java]: I think it should be "Return element clickable or not"

File: demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        }else check = false;
+        return check;        
+    }
+
+    //Declare element
+    private WebElement lblYourSearch = keyword.findElement(By.xpath("//span[contains(text(), 'Your Search')]"));
+    private WebElement lblSearchLocation = keyword.findElement(By.xpath("//div[@data-cmp = 'filterLocation']"));
+    private WebElement ddlDistance = keyword.findElement(By.id("2281868035"));
+    private WebElement txtZipcode = keyword.findElement(By.id("2225141853"));
+    private WebElement checkboxDelivery = keyword.findElement(By.xpath("//div[contains(text(), 'Include Extended Home Delivery')]"));
```
Comment (TrangDao25399) [inline demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java:90]: These WebElement should be declared at the top of class to make it clear and easier to find

File: demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleSearchLocationTest.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        basePage.navigateToPage("Sale");
+        CarForSaleSearchLocationPage carForSaleSearchLocationPage = new CarForSaleSearchLocationPage(driver);
+        Assert.assertTrue(carForSaleSearchLocationPage.isReturnTheResultOfDestanceMatchWithOptionWasSelected(distance_Miles));
+    }
+
+    @Test(testName = "verify That Return The Result Of Zip Code Should Be Matched With Option Was Selected", dataProvider = "zipCodeData", dataProviderClass = CarForSaleTestData.class, priority = 4)
+    public void verifyThatReturnTheResultOfZipCodeShouldBeMatchedWithOptionWasSelected(String zipCode) throws Exception{
+        BasePage basePage = new BasePage(driver);
+        basePage.navigateToPage("Sale");
+        CarForSaleSearchLocationPage carForSaleSearchLocationPage = new CarForSaleSearchLocationPage(driver);
```
Comment (TrangDao25399) [inline demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleSearchLocationTest.java:43]: In my opinion, We have 3 lines duplicated 3 times from the top. Maybe you can declare a function like actionOpenCarForSalePage or something else.

File: demo-automation-java/src/test/java/com/example/pages/BasePage.java (hunk 1 above)
Comment (tund-hcmue) [reply demo-automation-java/src/test/java/com/example/pages/BasePage.java]: Thanks for the comment ^-^

File: demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java (hunk 2 above)
Comment (tund-hcmue) [reply demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java:90]: Thanks for the comment, but it's ok ^--^

File: demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleSearchLocationTest.java (hunk 3 above)
Comment (tund-hcmue) [reply demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleSearchLocationTest.java:43]: Thanks for feedback, but it's also a function to open browser ^--^

**Comment by herrynhat** [comment]
Extend Report very good.

File: demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * Method verify that search location module should be worked correctly with option Zip Code
+     * @param zipcode: the zip code mentioned in distancedata.json file
+     * @return true or false
+     * @throws Exception
+     */
+    public boolean isReturnTheResultOfZipCodeMatchWithOptionWasSelected(String zipcode) throws Exception{
+        boolean check = false;
+        keyword.setText(txtZipcode, zipcode);
+        Thread.sleep(5000);
```
Comment (18521184) [inline demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java]: Can you replace sleep with something else

**Review by 18521184 — APPROVED** [review]

File: demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java (hunk 4 above)
Comment (tund-hcmue) [reply demo-automation-java/src/test/java/com/example/pages/carForSale/CarForSaleSearchLocationPage.java]: @18521184  Yeah, i will update it in the next PR, i need this PR merged before.

**Comment by tund-hcmue** [comment]
@18521184  I was resolved conflict, can you review it.

File: demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleFuelEconomyTest.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        CarForSaleFuelEconomyPage carForSaleFuelEconomyPage = new CarForSaleFuelEconomyPage(driver);
+        Assert.assertTrue(carForSaleFuelEconomyPage.isTheFuelEconomyFilterDisplayCorrect(fuel_Economy));
+    }
+
+    @Test(testName = "Verify That The Results Should Be Match With Chosen Fuel Economy Option", dataProvider = "fuelEconomyDataForTest", dataProviderClass = CarForSaleTestData.class)
+    public void verifyThatTheResultsShouldBeMatchWithChosenFuelEconomyOption(String fuel){
+        BasePage basePage = new BasePage(driver);
+        basePage.navigateToPage("Sale");
+        CarForSaleFuelEconomyPage carForSaleFuelEconomyPage = new CarForSaleFuelEconomyPage(driver);
+        carForSaleFuelEconomyPage.isReturnTheResultOfFuelEconomyMatchWithOptionWasSelected(fuel);
```
Comment (18521184) [inline demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleFuelEconomyTest.java]: ko có Assert gì hết à

**Review by 18521184 — COMMENTED** [review]
sao mấy cái test ko dùng Assert vậy

File: demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleFuelEconomyTest.java (hunk 5 above)
Comment (tund-hcmue) [reply demo-automation-java/src/test/java/com/example/tests/carForSale/CarForSaleFuelEconomyTest.java]: Thanks for feedback, I was update it.

**Comment by tund-hcmue** [comment]
Hi all, I just update script in my branch. Please review it.
@18521184 @ntmtrang 

1. Maintenance the test cases for module search location.
2. Implement the test cases for module search fuel economy and module search year.
- Create new file CarForSaleFuelEconomyPage.java, CarForSaleFuelEconomyTest, fueleconomydata.json, expectedfueleconomydata.json, expectedyeardata.json, yeardata.json, CarForSaleSearchYearModulePage.java, CarForSaleSearchYearModuleTest.java.
- Update file CarForSaleTestData.java 
- Create new keyword: waiForLocatorOfElementVisiable and findElementByLocator.

**Review by ntmtrang — APPROVED** [review]

**Pull request merged**